README.md
# Archeme

Archeme is a tool to describe and visualize schemes and diagrams. It’s designed primarily for architectural schemes.

Archeme may be used as a stand-alone command-line tool, or as a Python package.

To describe schemes and diagrams, Archeme provides YAML-based DSL (domain-specific language).

To draw them, Archeme uses [Graphviz](https://www.graphviz.org/). The engines `dot`, `neato`, and `fdp` are supported.

In fact, Archeme provides some new abstraction level over Graphviz. Unlike PlantUML, Archeme almost doesn’t limit your ability to use the flexible functionality of Graphviz. However, Archeme provides more convenient means to describe elements of the same types using classes, to specify the relative positions of elements, etc. Using YAML as the base of DSL makes it possible to work with diagram descriptions as with data structures—in Archeme itself and in applications integrated with it. So, Archeme can easily combine several scheme descriptions into one.

By using SVG as an output format, you may assign hyperlinks to scheme elements and thereby organize relationships between schemes—for example, in accordance with the [C4 methodology](https://c4model.com/).

## Installation

To install Archeme, run the command:

```bash
$ pip install archeme
```

After installation, the command `archeme` will be available in your system.

## CLI Usage

Currently Archeme provides 2 actions (commands):

* `generate`—to generate graph description that can be drawn with Graphviz from YAML-based DSL source;
* `merge`—to merge (combine, join) multiple DSL sources into a single DSL source.

### The `generate` Command Syntax

Generalized syntax of the `generate` command is the following:

```bash
$ archeme generate [-c|--config <config_file>] -i|--input <input_file> -o|--output <output_file> [-d|--draw <format>]
```

The `-i`/`--input` and `-o`/`--output` arguments are required, other arguments are optional.

The `-i`/`--input` parameter is used to specify YAML-based DSL source.

The `-o`/`--output` parameter points to the output file—graph description for Graphviz.

You may specify an optional config file by using the `-c`/`--config` argument. Both config and scheme description files use the same YAML-based DSL. It’s recommended to use a separate config file to specify some common settings that may be reused in multiple scheme descriptions of the same style. If both config file and scheme description set the parameter of the same name, the value specified in scheme description will override the value specified in config.

If the `-d`/`--draw` argument is given, Archeme will try to call Graphviz to draw an image from generated graph description, and the Graphviz calling command will be displayed. By default, Archeme doesn’t call Graphviz to visualize generated graph descriptions. The `-d`/`--draw` argument may take any output format value allowed by Graphviz (`png`, `svg`, etc.). By default, the `dot` engine is used. You may specify the necessary engine explicitly in DSL, see below.

### The `merge` Command Syntax

Generalized syntax of the `merge` command is more simple:

```bash
$ archeme merge -i|--input <input_file> -o|--output <output_file>
```

The command takes 2 mandatory arguments: `-i`/`--input` for input YAML-based DSL file, and `-o`/`--output` for output YAML-based DSL file. Output files of the `merge` command should be used as input files for the `generate` command. YAML-based DSL is described below.

## YAML-Based DSL

* See also: [Archeme DSL Quick Reference](https://github.com/foliant-docs/archeme/blob/master/dsl_reference.md)

To describe schemes and diagrams, YAML-based DSL is provided.

DSL, same as Graphviz, operates the graph theory terms: so, *graph* is the whole drawing; *node* is a functional component of a scheme or a diagram; *cluster* is a group of nodes that is usually highlighted visually; *edge* is a line that represents a relation between two nodes.

DSL may be used to:

* specify common settings of multiple elements;
* define scheme structure and relations between elements;
* describe each separate element—node, cluster, edge;
* set relative positions of elements;
* describe how it’s needed to combine multiple scheme descriptions into a single one.

### Common Setting

Common settings for multiple elements usually should be specified in the config file.

#### Graphviz Engine

The `engine` parameter sets the certain Graphviz engine to draw images. Archeme supports the `dot`, `neato`, and `fdp` engines. The default engine is `dot`.

Code example:

```yaml
engine: neato
```

You may read about the difference between engines in the [Graphviz documentation](https://www.graphviz.org/about/).

In relation to architectural schemes:

* `dot` is suitable for models with hierarchical relations; using `dot`, you can’t strictly control relative positions of elements—the positions depend of ranks of elements; `dot` supports clusters;
* `neato` allows to set strict fixed positions to each node; it doesn’t support clusters;
* `fdp` is similar to `neato` but it supports clusters; however, if clusters are used, `fdp` re-calculates the positions of nodes, they can’t be strictly fixed.

#### Settings Of Elements

In `elements` section, you may specify common settings of the element of different types—graph, node, cluster, edge. In the simplest case, DSL maps to Graphviz syntax. DSL code example:

```yaml
elements:
    graph:
        newrank: true
        rankdir: TB
    node:
        shape: box
        fixedsize: true
        width: 4
        height: 1
        style:
            - filled
            - rounded
    edge:
        dir: both
        arrowtail: dot
```

This example will be converted into the following description for Graphviz:

```
graph [newrank = "true", rankdir: "TB"]
node [shape = "box", fixedsize = "true", width = "4", height = "1", style = "filled, rounded"]
edge [dir = "both", arrowtail = "dot"]
```

All available parameters for graphs, nodes, edges are described in the [Graphviz documentation](https://www.graphviz.org/doc/info/attrs.html).

In addition to `graph`, `node`, and `edge` parameters, Archeme supports the analogous `cluster` parameter that allows to specify default settings for clusters. Code example:

```yaml
elements:
    cluster:
        labelloc: b
        labeljust: l
        shape: box
```

Note that you may reduce the number of lines of YAML-based DSL code. So, node settings from the example above may look like this:

```yaml
node {shape: box, fixedsize: true, width: 4, height: 1, style: [filled, rounded]}
```

Archeme allows to specify parameters that take lists of values, in different ways. The constructions `style: [filled, rounded]`, `style: filled, rounded`, `style: [filled]`, `style: filled` are valid.

Note that if a node, cluster, or edge has some custom `style` or another parameter with a list of values, it will **fully** override the default `style` or whatever. This is how it works in Graphviz, and Archeme doesn’t change this behavior for compatibility reasons.

For example, if the default `style` for nodes is `[filled, rounded]`, and the custom `style` of some node is `dashed`, only `dashed` will be applied to this node.

To control the settings of elements more flexible, classes are provided.

##### Classes Definition

You may define one or more classes for each type of elements (`node`, `cluster`, `edge`), and then assign any combination of defined classes to any separate element. Code example:

```yaml
elements:
    node:
        fixedsize: true
        penwidth: 3
        classes:
            generic:
                shape: box
                width: 4
                height: 1
                style: rounded
            database:
                shape: cylinder
                width: 3
                height: 3
            external_network:
                shape: circle
                width: 2.5
                height: 2.5
                style: filled
                fillcolor: '#f0f0f0'
```

In this example, 3 classes of nodes with the names `generic`, `database`, and `external_network` are defined. Note that using of classes doesn’t disallow to use global default settings.

##### Grid Settings

To control the relative positions of elements, Archeme provides the ability to describe nodes arrangement as a text grid. In the `elements` section of common settings you may define the horizontal and vertical intervals of the grid step. Code example:

```yaml
elements:
    grid:
        spacing_x: 400
        spacing_y: 250
```

The `spacing_x` parameter sets the horizontal interval and defaults to `500`; the `spacing_y` parameter sets the vertical interval and defaults to `200`. Grid spacing units are *points*; 1 point equals to 1/72 inch.

Note that `neato` and `fdp` engines use different units for the `pos` attribute: points and inches respectively. Archeme takes it into account. If you specify `engine: neato`, do not try to draw the resulting graph description with `fdp`, and vice versa. The `dot` engine doesn’t support positioning at all.

### Scheme Description

To describe a scheme or a diagram, it’s needed to:

* specify scheme structure—define nodes and clusters;
* specify relations between nodes—define edges;
* control the relative positions of nodes:
    * define a text grid, if the `neato` or `fdp` engine is used;
    * specify which nodes should have the same rank, if the `dot` engine is used.

#### Structure

To define nodes and clusters, Archeme DSL provides the `structure` section. You may assign Graphviz attributes and Archeme classes to any node and any cluster. For nodes, the `id` parameter is only required. All other parameters are optional. The `id` parameter is used to assign a text identifier to a certain node.

Simple code example:

```yaml
structure:
    - node:
        id: first
    - node:
        id: second
```

In this example, 2 nodes with the identifiers `first` and `second` are defined.

Note that the value of the `structure` parameter must be a list. If the `dot` engine is used, the order of nodes matters for their positioning. If the `neato` or `fdp` engine is used, the order of nodes doesn’t matter.

Clusters should be specified at the same level as nodes. Each cluster has nested `structure` that may include nodes and nested clusters. Code example:

```yaml
structure:
    - node:
        id: first
    - node:
        id: second
    - cluster:
        structure:
            - node:
                id: third
            - node:
                id: fourth
```

The following code example shows how it’s possible to assign Graphviz attributes and Archeme classes to nodes and clusters:

```yaml
structure:
    - node:
        id: first
        label: The First Node
        class:
            - amazing
            - awesome
    - cluster:
        label: The Wonderful Cluster
        style: filled
        fillcolor: '#cccccc'
        class: wonderful
        structure:
            ...
```

Single class name may be specified as a string; multiple class names should be specified as a list.

Archeme merges attributes of classes. Suppose 2 classes are defined as the following:

```yaml
classes:
    amazing:
        shape: box
        width: 4
        height: 1
        style:
            - rounded
            - filled
    awesome:
        height: 2
        style:
            - dashed
        fillcolor: '#99ccff'
```

If both styles are assigned to some node in the order: `[amazing, awesome]`, the resulting attributes assigned to the node will be:

```yaml
shape: box
width: 4
height: 2
style:
    - rounded
    - filled
    - dashed
fillcolor: '#99ccff'
```

So, the `style` attributes will be combined. Contradicting values of the attributes of the same names will be taken from the last class in the sequence (like `height` in this example).

#### Relations

To show relations between nodes, edges are used. To define a list of edges, use the `edges` section. Simple code example:

```yaml
edges:
    -   tail: first
        head: second
    -   tail: second
        head: third
```

In this example, 2 edges are defined. These edges specify relations between 3 nodes with the identifiers `first`, `second`, and `third`.

Archeme uses directed graphs only. So each edge should have a tail (“source,” “beginning”) and a head (“target,” “ending”). For example, if an edge represents interaction between a client and a server, it will be alright if a tail will be assigned to a client, and a head—to a server. However, client requests and server responses may be shown as two different counter-directed edges.

You may assign Graphviz attributes and classes to edges like in case with nodes and clusters. Example:

```yaml
edges:
    -   tail: first
        head: second
        label: HTTP API
        class: two_arrows
```

Archeme provides extended syntax for the `label` and `xlabel` attributes of edges. You may specify protocol and describe transferred data:

```yaml
    -   tail: first
        head: second
        label:
            protocol: HTTP
            data: Admin API
```

This code will be transformed into the following construction in a graph description:

```
"first" -> "second" [label = <<b>HTTP</b><br />Admin API>];
```

#### Positioning

Archeme implements a powerful concept to control relative positions of scheme elements. Positioning may be described as a multiline text grid. Suppose you have 10 nodes with the identifiers from `first` to `tenth`. Feel free to set their positions in this visual way:

```yaml
grid: |
    -        -    second    sixth
    first         third
                  fourth               eighth
                                       ninth
                  fifth     seventh
                                       -
                                       tenth
```

If the engine `neato` or `fdp` is used, the nodes will be strictly positioned with the `pos` attribute. Positions will be calculated using the values of the `elements.grid.spacing_x` and `elements.grid.spacing_y` parameters. A hyphen (`-`) means skipped step if it’s necessary not to place any nodes to the respective horizontal or vertical coordinate.

Grid steps are defined by start positions of node identifiers. So, the nodes `second`, `third`, `fourth`, and `fifth` from the example will be placed to the third horizontal step of the grid.

The `dot` engine ignores the `pos` attribute, but it allows to place nodes with the same rank to the same hierarchy level of the resulting layout. If the `dot` engine is used, Archeme analyzes a grid and defines groups (subgraphs) of nodes with the same ranks. For the example above and the `elements.graph.rankdir` attribute set to `LR` (left to right), Archeme will generate the following groups:

* `second`, `third`, `fourth`, `fifth`;
* `sixth`, `seventh`;
* `eighth`, `ninth`, `tenth`.

If the `elements.graph.rankdir` attribute is set to `TB` (top to bottom, default in Graphviz), for the example above Archeme will make the following groups:

* `second`, `sixth`;
* `first`, `third`;
* `fourth`, `eighth`;
* `fifth`, `seventh`.

If you don’t plan to use the engines `neato` and `fdp` to draw a certain scheme, and use only `dot`, you don’t need to describe grids. For `dot`, instead of using grids, you may define groups of nodes of the same ranks explicitly. Example of code that represents the last described case:

```yaml
groups:
    -   - second
        - sixth
    -   - first
        - third
    -   - fourth
        - eighth
    -   - fifth
        - seventh
```

### Combining Multiple Schemes

To combine multiple schemes or diagrams into a single one, the `merge` Archeme command is used. Archeme merges multiple scheme descriptions using the concept of modules.

A module is a scheme description that is used as a part of the structure of the resulting combined scheme description. Optionally some nodes of a certain module may be excluded from the resulting combined scheme description.

Each module should have an identifier. To avoid conflicts when some nodes have the same identifiers in different modules, node identifiers are combined with module identifiers in the resulting scheme description. For example, the node `api` from the module `backend` will get the identifier `backend.api` in the resulting combined scheme description.

Archeme works with modules that are described in separate files. To include some module into the resulting scheme structure, you should specify module identifier and module description file. Code example:

```yaml
structure:
    - module:
        id: backend
        file: path/to/backend.yml
    - module:
        id: frontend
        file: path/to/frontend.yml
```

It seems most convenient to describe modules in separate files, if Archeme is used as a stand-alone command-line tool. However, Archeme allows to specify a module description directly as a value of the `description` parameter that should be used instead of the `file` parameter. This way may be preferred when Archeme is used as a Python package in a third-party application.

In this example, 2 modules with the identifiers `backend` and `frontend` are added to the resulting structure.

If you need not to include some module nodes into the resulting structure, use the optional `exclude` parameter:

```yaml
structure:
    - module:
        id: frontend
        file: path/to/frontend.yml
        exclude:
            - client_device
            - cdn
```

In this example, the nodes with the identifiers `client_device` and `cdn` of the module `frontend` will not be taken into account in the resulting combined scheme description.

Feel free to add nodes, clusters, and edges to the resulting scheme description in the usual way:

```yaml
structure:
    - module:
        id: backend
        file: path/to/backend.yml
    - module:
        id: frontend
        file: path/to/frontend.yml
    - node:
        id: middleware
edges:
    -   tail: frontend.application
        head: middleware
    -   tail: middleware
        head: backend.api
```

Archeme provides smart merging of grids. Suppose you need to combine 4 modules with the identifiers `one`, `two`, `three`, and `four`. You may describe a grid that should be used in the resulting scheme description:

```yaml
grid: |
    one      two
    three    four
```

Archeme will generate the resulting grid in which, for example, the leftmost nodes of the module `two` will be placed at the same horizontal position as the leftmost nodes of the module `four`.

When merging modules, Archeme deeply processes the `structure`, `edges`, `grid`, and `groups` parameters of each module. Common parameters for multiple elements such as `engine` and `elements` will be merged using the simplest algorithm. It’s best to specify them in a separate config file that will not be involved in the merge process.

## Examples

Archeme repository includes [some examples](https://github.com/foliant-docs/archeme/tree/master/examples/) of architectural schemes.

Each example is located in its own folder that has the following structure:

* `source/*`—YAML-based DSL sources;
* `target/*`—generated graph descriptions for Graphviz, and resulting PNG images;
* `draw.sh`—shell script that executes the commands which create target files.

PNG images drawn with Graphviz are shown below.

### 1. Foliant Architecture, Simple, Drawn With `neato`

* [Directory with the example files](https://github.com/foliant-docs/archeme/tree/master/examples/01_foliant_architecture_simple/)

![](https://raw.githubusercontent.com/foliant-docs/archeme/master/examples/01_foliant_architecture_simple/target/architecture.png)

### 2. Foliant Architecture, Pretty, Drawn With `neato`

* [Directory with the example files](https://github.com/foliant-docs/archeme/tree/master/examples/02_foliant_architecture_pretty/)

![](https://raw.githubusercontent.com/foliant-docs/archeme/master/examples/02_foliant_architecture_pretty/target/architecture.png)


changelog.md
# 1.0.2

-   Fix bugs.

# 1.0.1

-   Allow to specify module descriptions directly, not in separate files. Add DSL quick reference.


archeme/generate.py
import re
from pathlib import Path
from collections import OrderedDict


class GenerateGraphvizSource():
    def _escape_value(self, source_value: str or int or float or bool) -> str:
        if isinstance(source_value, bool):
            if source_value:
                target_value = '"true"'

            else:
                target_value = '"false"'

        elif isinstance(source_value, str):
            target_value = source_value.replace('\r\n', '\n').replace('\n', '\\n').replace('"', '\\"')

            if not target_value.startswith('<') or not target_value.endswith('>'):
                target_value = f'"{target_value}"'

        else:
            target_value = f'"{source_value}"'

        return target_value

    def _serialize_single_line(self, entity_specification: str, entity_params: dict) -> str:
        serialized_entity_definition: str = f'{entity_specification} ['

        index: int
        key: str
        for index, key in enumerate(entity_params.keys()):
            if index > 0:
                serialized_entity_definition += ', '

            serialized_entity_definition += f'{key} = {self._escape_value(entity_params[key])}'

        serialized_entity_definition += '];\n'

        return serialized_entity_definition

    def _serialize_multiple_lines(self, params: dict) -> str:
        serialized_params: str = ''

        key: str
        for key in params.keys():
            serialized_params += f'{key} = {self._escape_value(params[key])};\n'

        return serialized_params

    def _csv_to_list(self, data: list or str) -> list:
        if isinstance(data, str):
            data = re.sub(r'\,\s*', ', ', data, flags=re.MULTILINE)
            data: list = data.split(', ')

        return data

    def _list_to_csv(self, data: str or list) -> str:
        if isinstance(data, list):
            data: str = ', '.join(data)

        return data

    def _merge_element_params(self, element_type: str, element_params: dict, dsl_full_scheme: dict) -> dict:
        element_styles: list = self._csv_to_list(element_params.pop('style', []))
        element_classes: list or str = element_params.pop('class', [])

        if not isinstance(element_classes, list):
            element_classes: list = [element_classes]

        element_type_params: dict = dsl_full_scheme.get('elements', {}).get(element_type, {}).copy()
        element_type_classes: dict = element_type_params.pop('classes', {})
        merged_classes_params: dict = {}

        class_name: str
        for class_name in element_classes:
            class_params: dict = element_type_classes.get(class_name, {})
            element_styles.extend(self._csv_to_list(class_params.get('style', [])))
            merged_classes_params.update(class_params)

        element_styles = sorted(list(set(element_styles)))

        if element_styles:
            element_params.update({'style': self._list_to_csv(element_styles)})

        merged_element_params = {**merged_classes_params, **element_params}

        if element_type == 'cluster':
            element_type_styles: list or str = element_type_params.pop('style', [])

            if element_type_styles:
                element_type_params['style'] = self._list_to_csv(self._csv_to_list(element_type_styles))

            merged_element_params = {**element_type_params, **merged_element_params}

        return merged_element_params

    def _get_nodes(self, dsl_full_scheme: dict) -> OrderedDict:
        def _get_substructure_nodes(substructure: list) -> OrderedDict:
            substructure_nodes: OrderedDict = {}

            substructure_item: dict
            for substructure_item in substructure:
                node_params: dict = substructure_item.get('node', {})
                cluster_params: dict = substructure_item.get('cluster', {})

                if node_params:
                    working_node_params: dict = node_params.copy()
                    node_id: str = working_node_params.pop('id', '')

                    if node_id:
                        working_node_params = self._merge_element_params(
                            'node',
                            working_node_params,
                            dsl_full_scheme
                        )

                        substructure_nodes[node_id] = working_node_params

                elif cluster_params:
                    substructure_nodes.update(_get_substructure_nodes(cluster_params.get('structure', [])))

            return substructure_nodes

        return {'nodes': _get_substructure_nodes(dsl_full_scheme.get('structure', []))}

    def _get_clusters(self, dsl_full_scheme: dict) -> dict:
        def _get_substructure_clusters(substructure: list, parent_cluster_id: str = '') -> dict:
            substructure_clusters: dict = {}
            clusters_count: int = 0
            nodes_count: int = 0

            substructure_item: dict
            for substructure_item in substructure:
                cluster_params: dict = substructure_item.get('cluster', {})
                node_params: dict = substructure_item.get('node', {})

                if cluster_params:
                    working_cluster_params: dict = cluster_params.copy()
                    cluster_structure: list = working_cluster_params.pop('structure', [])

                    working_cluster_params = self._merge_element_params(
                        'cluster',
                        working_cluster_params,
                        dsl_full_scheme
                    )

                    clusters_count += 1

                    cluster_id: str

                    if parent_cluster_id:
                        cluster_id = f'{parent_cluster_id}_{clusters_count}'

                    else:
                        cluster_id = f'{clusters_count}'

                    if clusters_count == 1:
                        substructure_clusters['clusters'] = {}

                    substructure_clusters['clusters'].update(
                        {cluster_id: _get_substructure_clusters(cluster_structure, cluster_id)}
                    )

                    substructure_clusters['clusters'][cluster_id].update(working_cluster_params)

                elif node_params:
                    if parent_cluster_id:
                        node_id: str = node_params.get('id', '')

                        if node_id:
                            nodes_count += 1

                            if nodes_count == 1:
                                substructure_clusters['nodes'] = []

                            substructure_clusters['nodes'].append(node_id)

            return substructure_clusters

        return _get_substructure_clusters(dsl_full_scheme.get('structure', []))

    def _get_edges(self, dsl_full_scheme: dict) -> dict:
        edges: list = []

        item: dict
        for item in dsl_full_scheme.get('edges', []):
            edge_params: dict = self._merge_element_params('edge', item, dsl_full_scheme)
            edges.append(edge_params)

        return {'edges': edges}

    def _get_subgraphs(self, dsl_full_scheme: dict) -> dict:
        def _get_grid_based_nodes_groups() -> list:
            grid_based_nodes_groups: list = []
            rankdir: str = dsl_full_scheme.get('elements', {}).get('graph', {}).get('rankdir', 'TB')
            grid_lines: list = dsl_full_scheme.get('grid', '').split('\n')
            nodes_lines: dict = {}
            nodes_starts: dict = {}

            index: int
            grid_line: str
            for index, grid_line in enumerate(grid_lines):
                nodes_matches = re.finditer(r'\S+', grid_line)

                for node_match in nodes_matches:
                    node_start: int = node_match.start(0)
                    node_id: str = node_match.group(0)

                    if not nodes_lines.get(index, None):
                        nodes_lines[index] = []

                    if not nodes_starts.get(node_start, None):
                        nodes_starts[node_start] = []

                    if node_id != '-':
                        nodes_lines[index].append(node_id)
                        nodes_starts[node_start].append(node_id)

            if rankdir == 'TB' or rankdir == 'BT':
                grid_based_nodes_groups = list(nodes_lines.values())

            elif rankdir == 'LR' or rankdir == 'RL':
                grid_based_nodes_groups = list(nodes_starts.values())

            return grid_based_nodes_groups

        return {'subgraphs': dsl_full_scheme.get('groups', []) + _get_grid_based_nodes_groups()}

    def _get_dsl_to_gv_intermediate(self, dsl_full_scheme: dict) -> dict:
        dsl_to_gv_intermediate: dict = {}

        for element_type in ('graph', 'node', 'edge'):
            element_type_params: dict = dsl_full_scheme.get('elements', {}).get(element_type, {}).copy()

            if element_type_params:
                element_type_params.pop('classes', None)

                element_type_styles: list or str = element_type_params.pop('style', [])

                if element_type_styles:
                    element_type_params['style'] = self._list_to_csv(self._csv_to_list(element_type_styles))

                dsl_to_gv_intermediate[element_type] = element_type_params

        def _append_positions(nodes: OrderedDict) -> OrderedDict:
            grid_lines: list = dsl_full_scheme.get('grid', '').split('\n')
            nodes_starts: dict = {}

            index: int
            grid_line: str
            for index, grid_line in enumerate(grid_lines):
                nodes_matches = re.finditer(r'\S+', grid_line)

                for node_match in nodes_matches:
                    node_start: int = node_match.start(0)

                    if not nodes_starts.get(node_start, None):
                        nodes_starts[node_start] = []

                    nodes_starts[node_start].append({'line': index, 'id': node_match.group(0)})

            grid_params = dsl_full_scheme.get('elements', {}).get('grid', {})
            spacing_x = grid_params.get('spacing_x', 500)
            spacing_y = grid_params.get('spacing_y', 200)
            node_position_x: int or float = 0

            if dsl_full_scheme.get('engine', 'dot') == 'fdp':
                spacing_x /= 72
                spacing_y /= 72

            node_start: int
            for node_start in sorted(nodes_starts.keys()):
                node: dict
                for node in nodes_starts[node_start]:
                    node_position_y: int or float = -node['line'] * spacing_y

                    if nodes.get('nodes', {}).get(node['id'], None) is not None:
                        if not nodes['nodes'][node['id']].get('pos', None):
                            if isinstance(node_position_x, float):
                                node_position_x = float('{0:.2f}'.format(node_position_x))

                            if isinstance(node_position_y, float):
                                node_position_y = float('{0:.2f}'.format(node_position_y))

                            nodes['nodes'][node['id']]['pos'] = f'{node_position_x}, {node_position_y}!'

                node_position_x += spacing_x

            return nodes

        if dsl_full_scheme.get('engine', 'dot') == 'neato' or dsl_full_scheme.get('engine', 'dot') == 'fdp':
            dsl_to_gv_intermediate.update(_append_positions(self._get_nodes(dsl_full_scheme)))

        else:
            dsl_to_gv_intermediate.update(self._get_nodes(dsl_full_scheme))

        dsl_to_gv_intermediate.update(
            {
                **self._get_clusters(dsl_full_scheme),
                **self._get_edges(dsl_full_scheme)
            }
        )

        if dsl_full_scheme.get('engine', 'dot') == 'dot':
            dsl_to_gv_intermediate.update(self._get_subgraphs(dsl_full_scheme))

        return dsl_to_gv_intermediate

    def generate_graphviz_source(self, dsl_full_scheme: dict) -> str:
        def _serialize_clusters(clusters: dict) -> str:
            serialized_clusters: str = ''

            cluster_id: str
            for cluster_id in clusters.keys():
                serialized_clusters += f'subgraph cluster_{cluster_id}' + ' {\n'

                cluster_params: dict = clusters[cluster_id].copy()
                nodes: list = cluster_params.pop('nodes', [])
                nested_clusters: dict = cluster_params.pop('clusters', {})

                serialized_clusters += self._serialize_multiple_lines(cluster_params)

                node_id: str
                for node_id in nodes:
                    serialized_clusters += f'"{node_id}";' + '\n'

                serialized_clusters += _serialize_clusters(nested_clusters)

                serialized_clusters += '}\n'

            return serialized_clusters

        dsl_to_gv_intermediate: dict = self._get_dsl_to_gv_intermediate(dsl_full_scheme)
        gv_output: str = 'digraph {\n'

        key: str
        for key in dsl_to_gv_intermediate.keys():
            if key in ('graph', 'node', 'edge'):
                gv_output += self._serialize_single_line(key, dsl_to_gv_intermediate[key])

            elif key == 'nodes':
                node_id: str
                for node_id in dsl_to_gv_intermediate[key].keys():
                    gv_output += self._serialize_single_line(f'"{node_id}"', dsl_to_gv_intermediate[key][node_id])

            elif key == 'clusters':
                gv_output += _serialize_clusters(dsl_to_gv_intermediate[key])

            elif key == 'edges':
                item: dict
                for item in dsl_to_gv_intermediate[key]:
                    edge_params: dict = item.copy()

                    edge_nodes: str = f'"{edge_params.pop("tail", "")}" -> "{edge_params.pop("head", "")}"'

                    if edge_params:
                        def _serialize_edge_label(edge_label: str or dict) -> str:
                            serialized_edge_label: str = ''

                            if isinstance(edge_label, dict):
                                serialized_edge_label = '<'

                                label_protocol: str = edge_label.get('protocol', '')

                                if label_protocol:
                                    serialized_edge_label += f'<b>{label_protocol}</b>'

                                label_data: str = edge_label.get('data', '')

                                if label_data:
                                    if label_protocol:
                                        serialized_edge_label += '<br />'

                                    serialized_edge_label += label_data

                                serialized_edge_label += '>'

                            elif isinstance(edge_label, str):
                                serialized_edge_label = edge_label

                            return serialized_edge_label

                        edge_label = edge_params.pop('label', '')

                        if edge_label:
                            edge_params['label'] = _serialize_edge_label(edge_label)

                        edge_xlabel = edge_params.pop('xlabel', '')

                        if edge_xlabel:
                            edge_params['xlabel'] = _serialize_edge_label(edge_xlabel)

                        gv_output += self._serialize_single_line(edge_nodes, edge_params)

                    else:
                        gv_output += edge_nodes + ';\n'

            elif key == 'subgraphs':
                subgraph: list
                for subgraph in dsl_to_gv_intermediate[key]:
                    gv_output += 'subgraph {\nrank = "same";\n';

                    node_id: str
                    for node_id in subgraph:
                        gv_output += f'"{node_id}";' + '\n'

                    gv_output += '}\n'

        gv_output += '}\n'

        return gv_output
